conda/plugins/virtual_packages/cuda.py

```diff
@@ -23,7 +25,63 @@ def cuda_version():
 
     Returns: version string (e.g., '9.2') or None if CUDA is not found.
     """
-    # platform-specific libcuda location
+    # Do not inherit file descriptors and handles from the parent process
+    # The `fork` start method should be considered unsafe as it can lead to
+    # crashes of the subprocess
+    ctx = mp.get_context("spawn")
+    queue = ctx.SimpleQueue()
```

Comment:
Is there a way to do this without starting a subprocess? This feels kind of brittle

Reply:
> Is there a way to do this without starting a subprocess?

@jakirkham Yes, another approach is to unset `"CUDA_VISIBLE_DEVICES"` in `os.environ` before calling `cuInit()` in the main process.

```python
# Empty `CUDA_VISIBLE_DEVICES` can cause `cuInit()` returns `CUDA_ERROR_NO_DEVICE`
# Invalid `CUDA_VISIBLE_DEVICES` can cause `cuInit()` returns `CUDA_ERROR_INVALID_DEVICE`
os.environ.pop("CUDA_VISIBLE_DEVICES", None)
```

The point here using subprocess is separate the `.so` handle (`fd`s) namespaces. Although it adds more complexity.

1. We never release the `.so` handle once we load `libcuda.so`. The `conda` program will always keep the NVIDIA device (`/dev/nvidia-uvm`) and kernel module (`nvidia_uvm`) in use while solving the environment. A benefit of using a subprocess here is we can release the `fd` as the subprocess ends. We use caches to memorize the result, so the subprocess will start only once.
2. Currently, `conda` is a standalone program, and it is not meant to be `import`ed. Unset and `CUDA_VISIBLE_DEVICES` and call `cuInit()` in the main process seems safe here. But if someone calls `cuInit()` with invalid `CUDA_VISIBLE_DEVICES` in advance. Then unsetting `CUDA_VISIBLE_DEVICES` and call `cuInit()` in conda internals will always fail. The `.so` handlers and `fd`s are messed up in the main process.